File: android/app/src/main/java/com/shelternav/app/viewmodel/MapViewModel.kt
package com.shelternav.app.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.shelternav.app.LatLon
import com.shelternav.app.ShelterInfo
import com.shelternav.app.bridge.ShelterNavJNI
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class MapViewModel : ViewModel() {
    private val _nearbyShelters = MutableStateFlow<List<ShelterInfo>>(emptyList())
    val nearbyShelters: StateFlow<List<ShelterInfo>> = _nearbyShelters.asStateFlow()

    private val _selectedShelter = MutableStateFlow<ShelterInfo?>(null)
    val selectedShelter: StateFlow<ShelterInfo?> = _selectedShelter.asStateFlow()

    private val _userLocation = MutableStateFlow<LatLon?>(null)
    val userLocation: StateFlow<LatLon?> = _userLocation.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    fun search(query: String) {
        // TODO: Filter shelters by name/address
    }

    fun findNearest() {
        val location = _userLocation.value ?: return
        viewModelScope.launch(Dispatchers.Default) {
            _isLoading.value = true
            _nearbyShelters.value = ShelterNavJNI.findNearest(
                lat = location.lat,
                lon = location.lon,
                radiusM = 5000,
                limit = 10,
            )
            _isLoading.value = false
        }
    }

    fun selectShelter(shelter: ShelterInfo) {
        _selectedShelter.value = shelter
    }

    fun updateUserLocation(lat: Double, lon: Double) {
        _userLocation.value = LatLon(lat, lon)
    }
}

File: ios/ShelterNav/ViewModels/MapViewModel.swift
import SwiftUI
@preconcurrency import CoreLocation

@MainActor
final class MapViewModel: NSObject, ObservableObject {
    @Published var nearbyShelters: [ShelterInfo] = []
    @Published var selectedShelter: ShelterInfo?
    @Published var userLocation: CLLocationCoordinate2D?
    @Published var isLoading = false
    @Published private(set) var recenterRequestID = UUID()

    private var allShelters: [ShelterInfo] = []
    private var activeQuery = ""
    private var hasLoadedInitialShelters = false

    private let bridge: ShelterNavBridge
    private let locationManager = CLLocationManager()

    init(bridge: ShelterNavBridge = .shared) {
        self.bridge = bridge
        super.init()

        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters
        locationManager.distanceFilter = 25

        if locationManager.authorizationStatus == .notDetermined {
            locationManager.requestWhenInUseAuthorization()
        } else {
            locationManager.startUpdatingLocation()
        }

        Task {
            await bridge.syncDatabase()
            if userLocation != nil {
                findNearest()
            }
        }
    }

    func search(query: String) {
        activeQuery = query.trimmingCharacters(in: .whitespacesAndNewlines)
        applySearchFilter()
    }

    func findNearest() {
        guard let location = userLocation else { return }
        Task {
            await loadNearest(for: location)
        }
    }

    func findNearestIfNeeded() {
        guard !hasLoadedInitialShelters, !isLoading else { return }
        findNearest()
    }

    func refreshData() async {
        guard let userLocation else { return }
        await bridge.syncDatabase()
        await loadNearest(for: userLocation)
    }

    func recenterOnUser() {
        guard userLocation != nil else { return }
        Haptics.itemSelected()
        recenterRequestID = UUID()
    }

    func selectShelter(_ shelter: ShelterInfo) {
        selectedShelter = shelter
        Haptics.pinSelected()
    }

    private func applySearchFilter() {
        guard !activeQuery.isEmpty else {
            nearbyShelters = allShelters
            return
        }

        nearbyShelters = allShelters.filter { shelter in
            shelter.name.localizedCaseInsensitiveContains(activeQuery) ||
                shelter.address.localizedCaseInsensitiveContains(activeQuery)
        }
    }

    private func loadNearest(for location: CLLocationCoordinate2D) async {
        isLoading = true
        defer { isLoading = false }

        let shelters = await bridge.findNearest(
            lat: location.latitude,
            lon: location.longitude,
            radiusM: 5000,
            limit: 10
        )
        allShelters = shelters
        applySearchFilter()
        hasLoadedInitialShelters = true
    }
}

extension MapViewModel: CLLocationManagerDelegate {
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        switch manager.authorizationStatus {
        case .authorizedWhenInUse, .authorizedAlways:
            manager.startUpdatingLocation()
        case .denied, .restricted:
            manager.stopUpdatingLocation()
        case .notDetermined:
            manager.requestWhenInUseAuthorization()
        @unknown default:
            manager.stopUpdatingLocation()
        }
    }

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let coordinate = locations.last?.coordinate else { return }
        userLocation = coordinate
        findNearestIfNeeded()
    }
}
